{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "082b4264",
   "metadata": {},
   "source": [
    "# Detecting an STR mutation in the *APC* gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a83155ce",
   "metadata": {},
   "source": [
    "Before working through this notebook, you should have completed the steps outlined in this repo's `README.md` file. i.e., you should have run GangSTR to generate a VCF file of STR genotypes for our sequencing sample.  \n",
    "\n",
    "In this notebook, we will programatically seach the VCF file to detect a STR mutation. Subsequently, we will try to interpret the possible effects of this mutation using the integrative genomics viewer (IGV) web app, Biopython, and Ensembl's variant effect predictor (VEP) tool.\n",
    "\n",
    "First, let's load some libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dae4cf3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from cyvcf2 import VCF\n",
    "from Bio import SeqIO, Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f606972",
   "metadata": {},
   "source": [
    "Our first task consists of searching for previously generated VCF file for the mutated STR locus. There should only be one mutation. There are many ways to approach this. I will outline two options here, feel free to pick the one that appeals most to you (or try both, or something completely different...). \n",
    "\n",
    "One option is a \"vanilla\" Python file parsing approach: open the VCF file, step through it line by line, and extract the VCF entry of interest. \n",
    "Some useful functions for this approach are:  \n",
    "\n",
    "```python\n",
    "open()\n",
    "str.startswith()\n",
    "str.strip()\n",
    "str.split()\n",
    "```\n",
    "\n",
    "Another approach is to use a dedicated VCF file parser that someone else developed. One example of such a library is [cyvcf2](https://brentp.github.io/cyvcf2/). Using the documentation for this library, you should be able to parse the VCF file and extract the entry containing the STR mutation.  \n",
    "\n",
    "*Note: In general, it is a good idea to use existing libraries for things like file parsing unless you have a very, very good reason not to. Existing libraries that are used by many people are typically very robust. Different people will have used them in many different setting, so many mistakes and bugs will have been noticed and fixed already.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "72cce82c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code goes here\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96435191",
   "metadata": {},
   "source": [
    "## Inspecting the mutation in the IGV web app"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdeab023",
   "metadata": {},
   "source": [
    "Open the [IGV web app](https://igv.org/app/) in your browser of choice. It just look like this:\n",
    "\n",
    "![](images/IGV_empty_example.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12073b67",
   "metadata": {},
   "source": [
    "By default, the human reference genome 'GRCh38' is loaded in the viewer. However, we want to use our custom APC reference sequence. Click on the `Genome` button in the top left, and select `Local File...`. You will then be prompted to select a file to upload. Navigate to the `data/alignments/` folder, and select `APC.fa` **AND** `APC.fa.fai` and the same time, and press `Open`, like so:\n",
    "![](images/IGV_upload_reference.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57dea9d0",
   "metadata": {},
   "source": [
    "The IGV web app should now display the reference sequence. You can now inspect the nucleotide sequence if you zoom in a bit, although this is not very interesting by itself:  \n",
    "![](images/IGV_just_APC.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4227c9cc",
   "metadata": {},
   "source": [
    "Let's add some of the other information we have available! Press the `Tracks` button at the top, choose `Local File...` again, and &mdash; similarly to how you uploaded the reference genome &mdash; add our sequence alignment (`APC_mut.bam`) + index (`APC_mut.bam.bai`) to the viewer. Then, do the same for the VCF file generated by GangSTR (this file does not need an index). Finally, upload the two files in `data/genome_annotation` (ending in `.gtf.gz` and `.gtf.gz.tbi`). This pair of files is are in [GTF](https://www.ensembl.org/info/website/upload/gff.html) format. This is a file format for annotation genomic sequences. This particular GTF file tells us where the transcript, exons, and coding sequences of the *APC* gene are located. After uploading all information, the final result should look something like what is shown below (once you zoom in a bit). Take a moment to look at the different tracks: do you know what each track represents?\n",
    "![](images/IGV_complete_example.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5247c6c9",
   "metadata": {},
   "source": [
    "Now that all our information is loaded into the genome viewer, it's time to take a look at the mutation in our sample. Navigate to the location of the STR mutation using the coordinates that you got from the GangSTR VCF file in the first part of this notebook (go to `Help`, `Documentation` in the IGV web app and look under `Navigation` if you're having trouble).  \n",
    "\n",
    "In what type of sequence does the STR mutation occur? Do you expect this mutation to have a functional impact on the protein transcribed from this gene? (especially pay attention to the genome annotation track, labelled `APC_canonical_relative_coordinates.gtf.gz`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cab3e8a8",
   "metadata": {},
   "source": [
    "## Manual assessment of STR mutation impact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f5866f8",
   "metadata": {},
   "source": [
    "Detecting the longest possible ORF in wild type and mutation sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "669cd631",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(141625 - 135054) \n",
    "135054 + divmod(141625 - 135054, 3)[0] * 3\n",
    "(141625 - 135054) % 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "14ee704b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "141624"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "135054 + divmod(141625 - 135054, 3)[0] * 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "219bb1e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def longest_orf(sequence: Seq, reverse_complement=False) -> Seq:\n",
    "    max_length = 0\n",
    "    for frame in (0, 1, 2):\n",
    "        frame_translation = sequence[frame:].translate()\n",
    "        stop_codon_idx = frame_translation.find(\"*\")\n",
    "        if stop_codon_idx == -1:\n",
    "            current_frame_length = len(frame_translation)\n",
    "        else:\n",
    "            current_frame_length = stop_codon_idx\n",
    "\n",
    "        if current_frame_length > max_length:\n",
    "            max_length = current_frame_length\n",
    "    return max_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "6131779d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mut = SeqIO.read(\"../../results/APC_mut_consensus.fa\", \"fasta\")\n",
    "wt = SeqIO.read(\"../../data/reference/APC.fa\", \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "a8a23fdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2190"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "longest_orf(wt.seq[135054:141625])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "57247406",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "817"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "longest_orf(mut.seq[135054:141625])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "40502525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# chr5:135,055-141,625\n",
    "wt.seq[135055:141625].translate().find(\"*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "bbb2e733",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.seq[135056:141625].translate().find(\"*\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "945feffe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2191"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(wt.seq[135055:141628].translate().find(\"*\"))\n",
    "len(wt.seq[135055:141628].translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "2a089d73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('QILRENNCLQTLLQHLKSHSLTIVSNACGTLWNLSARNPKDQEALWDMGAVSML...SV*')"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt.seq[135055:141628].translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "8a66b38c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "817\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2189"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(mut.seq[135055:141624].translate().find(\"*\"))\n",
    "len(mut.seq[135055:141624].translate())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "0eb58f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('QILRENNCLQTLLQHLKSHSLTIVSNACGTLWNLSARNPKDQEALWDMGAVSML...FKR')"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.seq[135055:141628].translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "21d3dabb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('AAAGCACCTACTGCTGAAAAGAGAGAGAGTGGACCTAAGC')"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt.seq[137461:137501]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "9427cd31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('KHLLLKRERVDLS')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wt.seq[137462:137501].translate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "df1e3ff8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('KHLLLKREWT*AS')"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut.seq[137462:137501].translate()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
